geode-core/src/upgradeTest/java/org/apache/geode/security/AuthExpirationDUnitTest.java

```diff
@@ -0,0 +1,109 @@
+import org.junit.Test;
+import org.junit.contrib.java.lang.system.RestoreSystemProperties;
+import org.junit.experimental.categories.Category;
+import org.junit.runner.RunWith;
+import org.junit.runners.Parameterized;
+
+import org.apache.geode.cache.Region;
+import org.apache.geode.cache.RegionShortcut;
+import org.apache.geode.cache.client.ClientCache;
+import org.apache.geode.cache.client.ClientRegionFactory;
+import org.apache.geode.cache.client.ClientRegionShortcut;
+import org.apache.geode.test.dunit.rules.ClientVM;
+import org.apache.geode.test.dunit.rules.ClusterStartupRule;
+import org.apache.geode.test.junit.categories.SecurityTest;
+import org.apache.geode.test.junit.rules.ServerStarterRule;
+import org.apache.geode.test.junit.runners.CategoryWithParameterizedRunnerFactory;
+
+@Category({SecurityTest.class})
+@RunWith(Parameterized.class)
+@Parameterized.UseParametersRunnerFactory(CategoryWithParameterizedRunnerFactory.class)
+public class AuthExpirationDUnitTest {
+
+  @Parameterized.Parameter
+  public String clientVersion;
+
+  @Parameterized.Parameters(name = "{0}")
+  public static Collection<String> data() {
+    // only test the current version and the latest released version
+    return Arrays.asList(CURRENT_VERSION, "1.13.3");
+  }
+
+  @Rule
+  public ClusterStartupRule lsRule = new ClusterStartupRule();
+
+
+  @Rule
+  public RestoreSystemProperties restore = new RestoreSystemProperties();
+
+  @Rule
+  public ServerStarterRule server = new ServerStarterRule()
+      .withProperty(SECURITY_MANAGER, ExpirableSecurityManager.class.getName())
+      .withRegion(RegionShortcut.REPLICATE, "region");
+
+  @After
+  public void after() {
+    // make sure after each test, the values of the ExpirationManager are reset
+    ExpirableSecurityManager securityManager =
+        (ExpirableSecurityManager) server.getCache().getSecurityService().getSecurityManager();
+    securityManager.reset();
+  }
+
+  @Test
+  public void clientShouldReAuthenticateWhenCredentialExpiredAndOperationSucceed()
+      throws Exception {
+    ExpirableSecurityManager.EXPIRE_AFTER = 10;
+    int serverPort = server.getPort();
+    ClientVM clientVM = lsRule.startClientVM(0, clientVersion,
+        c -> c.withCredential("data", "data")
+            .withPoolSubscription(true)
+            .withServerConnection(serverPort));
+
+    clientVM.invoke(() -> {
+      ClientCache clientCache = ClusterStartupRule.getClientCache();
+      ClientRegionFactory clientRegionFactory =
+          clientCache.createClientRegionFactory(ClientRegionShortcut.PROXY);
+      Region region = clientRegionFactory.create("region");
+      IntStream.range(0, 100).forEach(i -> {
+        region.put(i, "value" + i);
+      });
+    });
+
+    ExpirableSecurityManager securityManager =
+        (ExpirableSecurityManager) server.getCache().getSecurityService().getSecurityManager();
+
+    // assert that re-authentication happens
+    assertThat(securityManager.isExpired()).isTrue();
```

Comment:
It seems like this test has the potential to be flaky. This is assuming that expiration happened somewhere in the middle of the puts, based on a 10 millisecond timeout. Maybe it would be clearer just to have an API on the test security manager to *force* expiration and call that in the middle of the puts? I'm not sure this test needs to use timeouts at all.

Reply:
Good idea. The stressNewTest also catches the flakiness.